File: lib/haveapi/validators/confirmation.rb
module HaveAPI
  # Checks that two parameters are equal or not equal.
  #
  # Short form:
  #   string :param, confirm: :other_parameter
  #
  # Full form:
  #   string :param, confirm: {
  #     param: :other_parameter,
  #     equal: true/false,
  #     message: 'the error message'
  #   }
  #
  # +equal+ defaults to +true+.
  class Validators::Confirmation < Validator
    name :confirm
    takes :confirm

    def setup
      @param = simple? ? take : take(:param)
      @equal = take(:equal, true)
      @message = take(
        :message,
        @equal ? "must be the same as #{@param}"
               : "must be different from #{@param}"
      )
    end

    def describe
      {
          equal: @equal ? true : false,
          parameter: @param,
          message: @message,
      }
    end

    def valid?(v)
      if @equal
        v == params[@param]

      else
        v != params[@param]
      end
    end
  end
end

File: lib/haveapi/validator/confirmation.ex
defmodule HaveAPI.Validator.Confirmation do
  use HaveAPI.Validator

  def init(opts) do
    p = Keyword.fetch!(opts, :parameter)
    equal = Keyword.get(opts, :equal, true)

    %{
      parameter: p,
      equal: equal,
      message: opts[:message] || (if equal do
        "must be the same as #{p}"
      else
        "must be different from #{p}"
      end)
    }
  end

  def validate(%{equal: true} = opts, v, params) do
    return(opts, params[ opts[:parameter] ] === v)
  end

  def validate(%{equal: false} = opts, v, params) do
    return(opts, params[ opts[:parameter] ] !== v)
  end

  def return(_opts, true), do: :ok
  def return(opts, false), do: {:error, [opts.message]}
end
